src/Diagnostics.RuntimeHost/Controllers/DeploymentController.cs

```diff
@@ -0,0 +1,162 @@
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Diagnostics.RuntimeHost.Controllers
+{
+    [Authorize]
+    [Produces("application/json")]
+    [Route("api/deploy")]
+    public class DeploymentController : Controller
+    {
+
+        private IRepoClient devopsClient;
+        protected ICompilerHostClient _compilerHostClient;
+        protected IStorageService storageService;
+        private IInvokerCacheService detectorCache;
+
+        public DeploymentController(IServiceProvider services, IConfiguration configuration, IInvokerCacheService invokerCache)
+        {
+            this.storageService = (IStorageService)services.GetService(typeof(IStorageService));
+            this.devopsClient = (IRepoClient)services.GetService(typeof(IRepoClient));
+            this._compilerHostClient = (ICompilerHostClient)services.GetService(typeof(ICompilerHostClient));           
+            this.detectorCache = invokerCache;
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Post([FromBody] DeploymentParameters deploymentParameters)
+        {
+            // If all required parameters are empty, reject the request.
+            if(string.IsNullOrWhiteSpace(deploymentParameters.CommitId) && string.IsNullOrWhiteSpace(deploymentParameters.FromCommitId)
+                && string.IsNullOrWhiteSpace(deploymentParameters.ToCommitId) && string.IsNullOrWhiteSpace(deploymentParameters.StartDate)
+                && string.IsNullOrWhiteSpace(deploymentParameters.EndDate)) {
+                return BadRequest("The given deployment parameters are invalid");
+            }
+                     
+            DeploymentResponse response = new DeploymentResponse();
+            response.DeploymentGuid = Guid.NewGuid().ToString();
+            response.DeployedDetectors = new List<string>();
+            response.FailedDetectors = new Dictionary<string, string>();
+
+            var commitId = deploymentParameters.CommitId;
+            var timeTakenStopWatch = new Stopwatch();
+            timeTakenStopWatch.Start();
+            DiagnosticsETWProvider.Instance.LogRuntimeHostMessage($"Starting deployment operation for {response.DeploymentGuid}");
+            //  Get files to compile 
+            var filesToCompile = string.IsNullOrWhiteSpace(commitId) ? 
+                await this.devopsClient.GetFilesBetweenCommits(deploymentParameters)
+              : await this.devopsClient.GetFilesInCommit(commitId);
+
+            QueryResponse<DiagnosticApiResponse> queryRes = new QueryResponse<DiagnosticApiResponse>
+            {
+                InvocationOutput = new DiagnosticApiResponse()
+            };
+
+            IDictionary<string, string> references = new Dictionary<string, string>();
+
+
+            DiagnosticsETWProvider.Instance.LogRuntimeHostMessage($"Deployment id {response.DeploymentGuid}: {filesToCompile.Count} to compile");
+            //  Compile files
+            foreach( var file in filesToCompile)
```

Comment:
Maybe something you can consider in future to speed up this code (The slowness problem or maybe even throttling will surface for big detector repositories such as ours or AKS and a bunch  of edits by the team): 

- you can split the workload into 2 job tasks: The first task is to handle deleted files and the second task to handle compilation and publish. 
- For deleted files task, instead of updating storage one by one, batch them up and send one request to update.
- For compilation and publish, I see you are compiling each of the files sequentially and then publishing it. This code piece will be hit more regularly.  As a next iteration, you might want to add some degree of parallelism here as well.

What would be nice Rekha here is to keep a close watch (maybe add an email alert as well) on the latency of this API and send an email if latency exceeds a certain dangerous value (like 20 sec, reason - most http clients have 30 sec timeout). If we are seeing too many emails for this, it means we would need to optimize the speed of this API soonish.